openlibrary/accounts/model.py

```diff
@@ -597,42 +604,90 @@ def audit_accounts(email, password, test=False):
+            # If an Open Library account with a matching email account exist...
+            if ol_account:
+                # Check whether it is linked already, i.e. has an itemname
+                # set. We already determined that no OL account is
+                # linked to our IA account. Therefore this Open
+                # Library account having the same email as our IA
+                # account must have been linked to a different
+                # Internet Archive account.
+                if ol_account.itemname:
+                    return {'error': 'wrong_ia_account'}
+
+        # At this point, it must either be the case that (a)
+        # `ol_account` already links to our IA account (in which case
+        # `link` has a correct value), (b) that an unlinked
+        # `ol_account` shares the same email as our IA account and
+        # thus can and should be safely linked to our IA account, or
+        # (c) no `ol_account` which is linked or can be linked has
+        # been found and therefore, asuming
+        # lending.config_ia_auth_only is enabled, we need to create
+        # and link it.
+        if not ol_account:
+            if lending.config_ia_auth_only:
+                try:
+                    ol_account = OpenLibraryAccount.create(
+                        ia_account.itemname, email, password,
+                        displayname=ia_account.screenname,
+                        verified=True, test=test)
+                except ValueError as e:
+                    return {'error': 'max_retries_exceeded', 'msg': str(e)}
+
+                if not ol_account.itemname:
+                    ol_account.link(ia_account.itemname)
+                    stats.increment('ol.account.xauth.ia-auto-created-ol')
+
+                audit['link'] = ia_account.itemname
+                audit['has_ol'] = ol_account.email
+                web.ctx.conn.set_auth_token(ol_account.generate_login_code())
+
+        # So long as there's either a linked OL account, or an unlinked OL
         # account with the same email, set them as linked (and let the
-        # finalize logic link them, if necessary)
-        if ol_account:
+        # finalize logic link them, if needed)
+        else:
+            if lending.config_ia_auth_only:
+                if not ol_account.itemname:
+                    ol_account.link(ia_account.itemname)
+                    stats.increment('ol.account.xauth.auto-linked')
             if not ol_account.verified:
-                return {'error': 'account_not_verified'}
+                if lending.config_ia_auth_only:
+                    # The IA account is activated (verifying the
+                    # integrity of their email), so we make a judgement
+                    # call to safely activate them.
+                    ol_account.activate()
+                else:
+                    return {'error': 'ia_account_not_verified'}
             if ol_account.blocked:
```

Comment:
i.e. We already checked for blocked accounts up above, but you might comment that this is for the newly fetched OL account.